### PR TITLE
Fix Error: Couldn't import RGB Image: 00003002 (#373)

### PR DESCRIPTION
**Pull request by manteuffel723**
## Description
surface_descriptor_t::fds (graphics.h) did not get properly initialized in platf::kms::display_t::refresh.
That resulted in egs::import_source reading values from uninitialized variable/array elements from surface_descriptor_t and passing them to eglCreateImage.
eglCreateImage then fails with error code 3002 (EGL_BAD_ACCESS) (see https://registry.khronos.org/EGL/extensions/EXT/EGL_EXT_image_dma_buf_import.txt)

This fix ensures that surface_descriptor_t::fds does not contain uninitialized array elements after a successfull call to platf::kms::dislplay_t::refresh. Elements corresponding to invalid handles get set to -1.
That way the check that was already in place in egs::import_source `fd < 1` (line 490 graphics.h)  properly prevents passing invalid/unsupported/unwanted attributes to eglCreateImage. This check was basically dead code before this fix.

### Issues Fixed or Closed

Fixes Error: Couldn't import RGB Image: 00003002 #373`

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by KuleRucket** [comment]
Is it possible this fixes #453 as well?

File: src/platform/linux/kmsgrab.cpp [hunk 1]
```diff
@@ -583,6 +583,7 @@ class display_t : public platf::display_t {
 
     for(int y = 0; y < 4; ++y) {
       if(!fb->handles[y]) {
+        sd->fds[y] = -1;
```
Comment (ReenigneArcher) [inline src/platform/linux/kmsgrab.cpp:586]: Could you please add a comment above your change, describing what this is doing?

File: src/platform/linux/kmsgrab.cpp [hunk 2]
```diff
@@ -583,6 +583,7 @@ class display_t : public platf::display_t {
...
         // It's not clear wheter there could still be valid handles left.
```
Comment (ReenigneArcher) [inline src/platform/linux/kmsgrab.cpp:587]: I know this wasn't your change, but since you're changing this anyway, can you fix `whether`?

**Comment by Deytron** [comment]
I just tested the fix, and I can confirm that this fixes both #373 and #453 (00003002 and 00003004 respectively)

## Test environment

OS: EndeavourOS Linux x86_64 
Kernel: 6.1.4-zen2-1-zen 
Resolution: 1920x1080 
DE: Plasma 5.26.5 
WM: kwin-wayland
CPU: AMD Ryzen 5 3600X (12) @ 4.000GHz 
GPU: NVIDIA GeForce RTX 2060 Rev. A 
Memory: 32GB

### What works

- NVENC HEVC encoder seems to work
- Mouse and keyboard both work
- Latency is close to zero in LAN
- Sound works

### What does not work

- May be me, but the web UI has a blank page and gives me (localhost : not authorized)
- I get a black screen on the client
- This errors pops up in terminal :

`[2023:01:13:11:39:24]: Error: GL: /home/romain/Sunshine/src/platform/linux/graphics.cpp:529: [00000501]`

Line 529 only contains `gl_drain_errors;` , so I guess the error lies somewhere else

Interestingly, Sunshine tries to capture an xrandr window even on Wayland

**Comment by KuleRucket** [comment]
This fixes #453 for me too. For the first time ever I can screencast with KMS under wayland on Nvidia

**Comment by KuleRucket** [comment]
The graphics.cpp Error: GL: problem is reported in #404

**Comment by manteuffel723** [comment]
> The graphics.cpp Error: GL: problem is reported in #404

What error do you get? 00000501 (GL_INVALID_VALUE) like Deytron or 00000502 (GL_INVALID_OPERATION) as in #404?

Looks like calls to:
- BindTexture
- GenTextures
   
can cause these errors. Conditions for these Errors are documented in https://registry.khronos.org/OpenGL-Refpages/gl4/.
Maybe you can do some printf debugging to see, where exactly the error comes from.  There are 2 calls to BindTexture in import_source. I didn´t find documentation for EGLImageTargetTexture2DOES, so maybe even after that call.
Additionally there are calls to GenTextures and BindTexture in tex_t::make (line 26 graphics.cpp) that is called from import_source.

I don´t have a nvidia gpu, so i cannot reprduce it. Screencasting works just fine on my amd card now.

**Comment by KuleRucket** [comment]
> What error do you get? 00000501 (GL_INVALID_VALUE) like Deytron or 00000502 (GL_INVALID_OPERATION) as in #404?

You're right I didn't notice the difference. I get 00000501 like Deytron above.

**Comment by Deytron** [comment]
I will be able to investigate in the following week, I'll keep you updated

**Comment by zastrixarundell** [comment]
Compiling this on fedora silverblue 37 and running under an Ubuntu 22.04 container I managed to not get #453 but the screen on the client was black with my cursor. I did get `Screencasting with X11` rather than KMS.

This issue might be just on my side because of the container mess as I can't install on my system because of #210.

**Comment by Deytron** [comment]
I don't know if this will help much, but here's some debugging I did :  

```
[2023:01:22:22:18:31]: Error: GL: /home/romain/Sunshine/src/platform/linux/graphics.cpp:534: [00000501]

eglCreateImage returned 0x7f1378379c91

egl_display: 0x7f137800b4f0
attribs: 0x7f139d7f9910
rgb->tex[0]: 1
rgb->xrgb8: 0x7f1378379c91
rgb->tex[0]: 1
rgb->xrgb8: 0x7f1378379c91
```

**Comment by Deytron** [comment]
For reference, #649 should have fixed the GL errors, but it didn't fix it on my end

**Comment by zastrixarundell** [comment]
Another thing I noticed with a multi-monitor setup (3 monitors horizontally aligned) is that when I move the cursor on the monitor at `+0+0` position I get a weird mouse-following artifact. Keep in mind that after I connected on moonlight I did NOT move the mouse of the client but rather I moved it on the host. 

So the video feed manages to capture the cursor but not anything below it.

I had a similar issue when I installed openbox without a compositor on a clean fedora installation.

https://user-images.githubusercontent.com/30195425/214153681-a2ce9f8e-9730-42c3-a9ab-2eafde38629c.mp4

**Comment by ReenigneArcher** [comment]
Am I correct in that this solves 00003002, and only removes the error for 00003004?

**Comment by zastrixarundell** [comment]
I can't say for 00003002 but on a wayland + nvidia setup it does remove the error 00003004 and for the first time I can initialize the stream. The video stream itself doesn't work how it should but everything else does (audio and input capture).

**Comment by Deytron** [comment]
> Am I correct in that this solves 00003002, and only removes the error for 00003004?

Yes, this is correct, both 0003002 and 00003004 are solved. I guess we should open a new issue concerning the new GL error.

**Comment by Deytron** [comment]
Created issue #838 for further investigation and better tracking

**Comment by KuleRucket** [comment]
I think this fix should be merged. At least the behaviour is better than it was.

**Pull request closed without merging**